modules/relay/src/main/RelayPush.scala

```diff
@@ -29,46 +29,57 @@ final class RelayPush(sync: RelaySync, api: RelayApi, irc: lila.irc.IrcApi)(usin
     if rt.round.sync.hasUpstream
     then fuccess(Left(LilaInvalid("The relay has an upstream URL, and cannot be pushed to.")))
     else
-      rt.round.sync.nonEmptyDelay.fold(push(rt, pgn)): delay =>
-        validateForDelay(rt, pgn) match
-          case Some(err) => fuccess(Left(err))
-          case None      => after(delay.value.seconds)(push(rt, pgn))
+      pgnToGames(rt, pgn) match
+        case Left(err) => fuccess(Left(err))
+        case Right(games) =>
+          rt.round.sync.nonEmptyDelay.fold(push(rt, games)): delay =>
+            after(delay.value.seconds)(push(rt, games))
 
-  private def validateForDelay(rt: RelayRound.WithTour, pgn: PgnStr): Option[LilaInvalid] =
+  private def push(rt: RelayRound.WithTour, games: Vector[RelayGame]): Fu[Result] =
+    workQueue(rt.round.id):
+      sync
+        .updateStudyChapters(rt, games)
+        .map: res =>
+          SyncLog.event(res.nbMoves, none)
+        .recover:
+          case e: Exception => SyncLog.event(0, e.some)
+        .flatMap: event =>
+          if !rt.round.hasStarted && !rt.tour.official && event.hasMoves then
+            irc.broadcastStart(rt.round.id, rt.fullName)
+          api
+            .update(rt.round): r1 =>
+              val r2 = r1.withSync(_ addLog event)
+              val r3 = if event.hasMoves then r2.ensureStarted.resume else r2
+              r3.copy(finished = games.nonEmpty && games.forall(_.ending.isDefined))
+            .inject:
+              event.error.fold(Right(event.moves))(err => Left(LilaInvalid(err)))
+
+  private def pgnToGames(rt: RelayRound.WithTour, pgnBody: PgnStr): Either[LilaInvalid, Vector[RelayGame]] =
     Reader
-      .full(pgn)
+      .full(pgnBody)
       .fold(
-        err => LilaInvalid(err.value).some,
+        err => Left(LilaInvalid(err.value)),
         _ match
-          case Reader.Result.Incomplete(_, errorStr) => LilaInvalid(errorStr.value).some
+          case Reader.Result.Incomplete(_, errorStr) => Left(LilaInvalid(errorStr.value))
```

Comment:
I am worried that dgt sending kings to center as actual moves will fail here and we'll never get to fixDgtKingsInTheCenter. But I'm not sure how this looks and could use some help to test.